### PR TITLE
build: Overlay build vars expand ${ZEPHYR_<module>_MODULE_DIR}

### LINKED ISSUE #41830: CONF_FILE, OVERLAY_CONFIG parsing expands ${ZEPHYR_<whatever>_MODULE_DIR}
**Is your enhancement proposal related to a problem? Please describe.**
Downstream project integrators may be pulling content in from multiple downstream repositories, and may be configuring multiple similar projects. In order to reference config files in another module, integrators must explicitly provide either absolute pathnames, project-dir-relative pathnames, or ZEPHYR_BASE-prefixed pathnames to each file. This solution does not reflect that the mounting points of ZEPHYR_MODULES is arbitrary and is difficult to maintain when an imported west manifest changes the path to a module.

**Describe the solution you'd like**
I would like to see the parsing of the CMake variables CONF_FILE, OVERLAY_CONFIG expand ${ZEPHYR_MYMODULE_MODULE_DIR} substrings into the absolute path to the referenced module.  This enables integrators to reference overlay files in a different module without caring where that module is located in the workspace.

It seems this can be easily implemented where needed similar to the following example:
Replace: 
    `string(REPLACE " " ";" OVERLAY_CONFIG_AS_LIST "${OVERLAY_CONFIG}")`
With:
    `string(CONFIGURE "${OVERLAY_CONFIG}" OVERLAY_CONFIG_EXPANDED)`
    `string(REPLACE " " ";" OVERLAY_CONFIG_AS_LIST "${OVERLAY_CONFIG_EXPANDED}")`

As of Zephyr 2.7.0 it appears this replacement is only needed in `cmake/app/boilerplate.cmake` and `cmake/kconfig.cmake`.

**Describe alternatives you've considered**
I have not considered any other options.

**Additional context**

### PR DESCRIPTION
**Pull request by gregshue**
Support referencing module directories by name in CONF_FILE,
OVERLAY_CONFIG, and DTC_OVERLAY_FILE so that projects can reference
overlay files in arbitrary modules.

Fixes #41830

Signed-off-by: Gregory Shue <gregory.shue@legrand.us>

### CODE REVIEW

**Review by de-nordic — DISMISSED** [review]

File: doc/develop/application/index.rst [hunk 1]
```diff
@@ -474,7 +474,20 @@ Application Configuration Directory
 Zephyr will use configuration files from the application's configuration
 directory except for files with an absolute path provided by the arguments
 described earlier, for example ``CONF_FILE``, ``OVERLAY_CONFIG``, and
-``DTC_OVERLAY_FILE``.
+``DTC_OVERLAY_FILE``.  The absolute path to a file in a module will be
+generated by referring to it in your project's :file:`CMakeLists.txt` as:
+
+.. code-block:: cmake
+
+   set(CONF_FILE "\${ZEPHYR_MODULENAME_MODULE_DIR}/<module-relative-pathname>")
```
Comment (tejlmand) [inline doc/develop/application/index.rst:482]: Could you please push the correct version ?

This is old version where changes was requested here:
https://github.com/zephyrproject-rtos/zephyr/pull/42057#discussion_r807340997

**note** you most likely need to manually expand collapsed comments in githug for the link to take you to the right comment.
Old comment copy:
> Looking at the rendered output, I feel this particular feature fills to much, compared to the description of the other parts:
image
>
> How about just a short mentioning of this possibility, for example:
```suggestion
``DTC_OVERLAY_FILE``. A file in a Zephyr module can be referred by escaping the Zephyr module dir variable like this ``\${ZEPHYR_<module>_MODULE_DIR}/<path-to>/<file>`` when setting any of said variables.
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please push the correct and approved version.

Current version is older with unaddressed comments.

**Comment by gregshue** [comment]
@tejlmand Thank you for pointing out my mistake.  I have corrected the changes that I could map to the current files.  Some of the relevant documentation has been restructured.  I will need some guidance on further documentation updates.

**Review by tejlmand — DISMISSED** [review]
lgtm

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by de-nordic — DISMISSED** [review]

**Comment by gregshue** [comment]
Rebased, updated test source to new ZTest API.

**Review by tejlmand — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**